[Step3] 로또 (자동)

soob-forest opened the pull request:
아직 모자란 것 같지만, 
코멘트를 한 번 받고 진행하는게 더 좋아보여 리뷰 요청 드립니다.

### 도메인
- Money : value와 unit을 이용해 돈을 나타내기 위한 클래스

- LottoNumber : 1 - 45 사이의 숫자를 포장 
- Lotto : 로또에 관련된 역할을 위한 클랙스
- Lottos : Lotto 리스트를 가지는 일급 콜렉션

- WinningNumbers : 당첨 번호와 관련된 역할을 위한 클래스

- RandomNumberMachine : 랜덤한 LottoNumber을 생성을 위한 클래스

- AutoLottoIssuer : 주문 가격에 맞는 로또를 발행 하는 역할

- LottoStatistic : 주문한 로또 리스트와 당첨 번호를 이용한 통계 관련 역할을 하는 클래스

- MatchResult : 로또의 매치 결과를 나타내기 위한 Enum

soob-forest committed:
docs: 로또 자동 기능 정리

soob-forest committed:
feat(LottoNumber): 로또의 숫자 원시값을 포장하기 위한 클래스

- 로또 숫자 범위인지 validate 함수 추가

soob-forest committed:
feat(Money): 돈을 나타내기 위한 클래스

- 값과 단위를 가짐
- 음수 validate
- 나누기 멤버함수 추가

soob-forest committed:
feat(MatchResult.enum): 로또 매치 결과를 위한 enum

- 동일한 숫자 개수에 따른 상금을 가지도록 함

soob-forest committed:
feat(Lotto): 로또 클래스 생성

- 로또 숫자 개수, 중복 숫자 validate 추가
- 매치 결과 반환 멤버함수 추가

soob-forest committed:
feat(RandomNumberMachine): 랜덤 로또 숫자를 생성하기 위한 클래스

- 1~45 범위의 LottoNumber 타입의 Stack을 관리
- 생성자 호출 시, LottoNumber Stack을 shuffle
- popRandomNumber 멤버한수 구현
- pop 한 RandomNumber를 refill한 후, shuffle하기 위한 멤버함수 구현

soob-forest committed:
feat(LottoRandomFactory): 랜덤 로또 1개를 생성하기 위한 클래스

soob-forest committed:
feat(AutoLottoIssuer): 로또 자동 발행 하기 위한 클래스 추가

- 주문가격에 따라 로또를 자동생성

soob-forest committed:
docs: 요구사항 정리

soob-forest committed:
docs: 로또 리스트 일급 컬렉션 내용 추가

soob-forest committed:
feat(Lottos): 로또 리스트 일급 컬렉션 추가

- 매치 결과에 해당하는 로또 리스트를 반환하기 위한 멤버함수 추가

soob-forest committed:
feat(Money): Money에 더하기 추가

soob-forest committed:
refactor(Money): Money의 value타입 double로 변경

soob-forest committed:
refactor(Lotto): Lotto의 가격 반환 함수 추가

soob-forest committed:
feat(Lottos): Lottos의 lottoList의 총 금액 반환 함수 추가

soob-forest committed:
feat(LottoStatistic): Lottos의 통계를 위한 도메인 클래스 추가

- 수익률 계산
- 로또 당첨 결과 계산

soob-forest committed:
feat(LottoNumber): toString 정의

soob-forest committed:
feat(Lotto): toString 정의

soob-forest committed:
feat(Lottos): toString 정의

soob-forest committed:
feat(Money): toString 정의

soob-forest committed:
feat(LottoStatistic): toString 정의

soob-forest committed:
refactor(AutoLottoIssuer): 발행 시 Lottos를 반환하도록 수정

soob-forest committed:
feat(MatchResult): MatchResult의 matchCount getter 추가

soob-forest committed:
feat(LottoStatistic): 스트링 배열 타입의 prizeNumbers로도 생성 할 수 있도록 변경

soob-forest committed:
refactor(Money): 스트링 타입의 value로 생성 가능하도록 변경

soob-forest committed:
feat(View): 사용자와 interaction 계층 추가

soob-forest committed:
feat(LottoController): 컨트롤러 추가

soob-forest committed:
feat(Application): app entry 포인트 추가

soob-forest committed:
refactor(LottoStatistic): 각 matchResult의 prize를 계산하는 로직을 함수로 생성

soob-forest committed:
feat(WinningNumbers): 당첨번호에 관한 책임을 갖는 클래스 생성

soob-forest committed:
refactor(WinningNumbers): WinningNumbers를 이용하도록 변경

chae-yh commented on src/main/java/lotto/Application.java:
```diff
@@ -0,0 +1,22 @@
+
+public class Application {
+    static final int LOTTO_NUMBER_SIZE = 6;
+    static final int LOTTO_PRICE = 1000;
```
저희는 특정 기능을 도메인의 책임과 역할에 맞게 나누고 있습니다. lotto의 크기 및 가격은 프로그램을 실행하는 해당 클래스에 어울리는 상수는 아닐 것 같습니다. 적절한 클래스로 옮겨보는 것은 어떨까요...?

soob-forest replied:
도메인 외부의 값들을 이용해 로또가 가져야 할 값을 조절해 
다양한 로또의 형태를 가질 수 있도록 구현을 하려다 보니, 본질에서 벗어 나버린 것 같습니다..

chae-yh commented on src/main/java/lotto/Application.java:
```diff
@@ -0,0 +1,22 @@
+
+    public static void main(String[] args) {
+
+        AutoLottoIssuer autoLottoIssuer = new AutoLottoIssuer(
```
로또를 만드는 것은 더 안쪽(예를 들면 컨트롤러 등)에서 해줄 수 있지 않을까요...?

soob-forest replied:
컨트롤러로 옮겼습니다.

chae-yh commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,42 @@
+        this.autoLottoIssuer = autoLottoIssuer;
+    }
+
+    public void start() {
```
한 메서드가 너무 긴데 조금 나눠볼 수 있지 않을까요...?

soob-forest replied:
함수를 역할별로 나누는 방향으로 변경했습니다.

chae-yh commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.List;
+import java.util.Set;
+
+public class Lotto {
```
원시값을 포장한 클래스인데, 필드가 너무 많습니다. LottoNumbers 만 가지고 있으면 되지 않을까요...?

soob-forest replied:
LottoNumbers  만 가지도록 변경했습니다.

LOTTO_NUMBER_SIZE와 LOTTO_PRICE는 Lotto에 필요해 보여 static 변수로 가지고 있도록 했습니다.

chae-yh commented on src/main/java/lotto/domain/MatchResult.java:
```diff
@@ -0,0 +1,51 @@
+        this.cashPrize = cashPrize;
+    }
+
+    public static MatchResult from(int matchCount) {
```
이렇게 if절로 만들기보다 더욱 간단하게 확인할 수는 없을까요...?

soob-forest replied:
Enum의 values 를 이용하도록 변경했습니다.

chae-yh commented on src/main/java/lotto/domain/AutoLottoIssuer.java:
```diff
@@ -0,0 +1,33 @@
+
+
+    private void validate(double lottoCount) {
+        if (!(lottoCount == Math.floor(lottoCount)) || Double.isInfinite(lottoCount)) {
```
부동 소수점 연산의 경우는 숫자가 정확히 나오지 않을 수가 있습니다. 그래서 이 부분 로직에서 의도하지 않은 결과가 발생할 수 있을 것 같아요.
정확한 데이터가 필요하다면 BigDecimal을 쓰는 게 좋을 것 같습니다 : )

soob-forest replied:
BigDecimal 로 Money의 value를 변경했습니다.

chae-yh commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,59 @@
+
+    private LottoNumber(int lottoNumber) {
+        this.lottoNumber = lottoNumber;
+        validate();
```
validate를 먼저하고 값을 할당하는 것은 어떨까요...?

soob-forest replied:
validate를 먼저 하도록 변경했습니다.

chae-yh commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    private void validate() {
+        if (this.lottoNumber < 1 || this.lottoNumber > 45) {
```
매직넘버는 상수로 추출하는 것은 어떨까요...?

soob-forest replied:
자꾸 빼먹네요 상수로 추출 하는 것을ㅜ
변경했습니다.

chae-yh commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Override
+    public int compareTo(LottoNumber o) {
```
Integer.compare() 를 써봐도 좋을 것 같습니다 ㅎㅎ

soob-forest replied:
Integer.compare() 을 사용하도록 했습니다.

chae-yh commented on src/main/java/lotto/domain/LottoRandomFactory.java:
```diff
@@ -0,0 +1,23 @@
+package lotto.domain;
+
+public class LottoRandomFactory {
```
여기도 너무 많은 필드를 가지고 있습니다. 파라미터로 받거나 지역변수로 처리하면 어떨까요...?

soob-forest replied:
필요한 파라미터만 남겼습니다.

파라미터가 많으면 안좋은 점이 뭐가 있을까요??
해당 클래스의 책임이나 역할이 불분명해 지는 이유일까요?

chae-yh replied:
말씀하신 책임과 역할이 불분명해지는 사유도 있습니다. 예를 들어 Lotto라는 일급컬렉션을 보면 저는 int set 혹은 LottoNumber set이 있겠구나..라고 기대하고 코드를 읽겠지만, 다른 변수들이 존재한다면 해당 클래스의 코드를 파악하는데 어려움이 있을 것 같습니다.
두번째로, 전역 변수의 경우 값을 중간에 바꾸기가 쉽고, 그럴 경우 사용자가 해당 데이터를 사용하기가 쉽지 않은 이유도 있을 것 같습니다. (물론 final로 막을 수는 있겠습니다만...ㅎ.ㅎ;;)
마지막으로, 저희는 클린 코드를 연습하는 입장이라는 점도 있을 것 같습니다! 원시값을 wrapping하고, 일급컬렉션을 만드는 과정에서 극단적으로 작업해보는 것이 도움이 될 것 같습니다 : )

chae-yh commented on src/main/java/lotto/domain/RandomNumberMachine.java:
```diff
@@ -0,0 +1,34 @@
+        Collections.shuffle(numbers);
+    }
+
+    public LottoNumber popRandomNumber() {
```
이 방법은 재밌네요 👍

chae-yh commented on src/main/java/lotto/domain/LottoStatistic.java:
```diff
@@ -0,0 +1,103 @@
+
+
+    @Override
+    public String toString() {
```
LottoStatistic은 도메인 객체이므로 이 부분처럼 view 관련 로직을 가질 필요는 없을 것 같습니다!

soob-forest replied:
view 관련 로직 제거했습니다.

chae-yh commented on src/main/java/lotto/domain/LottoStatistic.java:
```diff
@@ -0,0 +1,103 @@
+        return builder.toString();
+    }
+
+    private String toMatchResultString(MatchResult matchResult) {
```
이 부분도 데이터만 전달해주고, 실제 출력하는 것은 view로 보내면 좋지 않을까요...?

soob-forest replied:
view 관련 로직 제거했습니다.

chae-yh commented on src/main/java/lotto/domain/LottoStatistic.java:
```diff
@@ -0,0 +1,103 @@
+                matchResult) + "개\n";
+    }
+
+    private String toEarningString() {
```
이 부분도 view로 가면 좋을 것 같습니다!

soob-forest replied:
view 관련 로직 제거했습니다.

chae-yh commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,57 @@
+public class Lotto {
+    private static final int DEFAULT_NUMBER_SIZE = 6;
+    private final Money price;
+    private final List<LottoNumber> lottoNumbers;
```
여기는 중복을 허용하지 않는 자료구조를 사용하면 어떨까요...?

soob-forest replied:
validate에서 중복 데이터를 걸러내기 때문에 필요없다고 생각했는데,
명확한 자료구조를 사용하는게 좋을 거 같아 변경했습니다.

chae-yh commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,46 @@
+        this.lottoList = new ArrayList<>(lottoList);
+    }
+
+    public static Lottos from(List<Lotto> lottoList) {
```
static factory method 좋습니다 👍

chae-yh commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,46 @@
+                result.add(lotto);
+            }
+        }
+        return Collections.unmodifiableList(result);
```
불변객체로 return해주는 것 좋네요 👍

chae-yh commented on src/main/java/lotto/domain/LottoStatistic.java:
```diff
@@ -0,0 +1,103 @@
+import java.util.Map;
+import java.util.Map.Entry;
+
+public class LottoStatistic {
```
여기도 필드가 너무 많네요 lottos와 winningNumbers는 없어도 되지 않을까요...?

soob-forest replied:
필요한 멤버 변수만 남겨두고 제거했습니다.

chae-yh commented on src/main/java/lotto/domain/MatchResult.java:
```diff
@@ -0,0 +1,51 @@
+package lotto.domain;
+
+public enum MatchResult {
```
순위 값을 enum으로 잘 정의해주셨습니다!
테스트 코드도 추가 되면 좋을 것 같아요 : )

soob-forest replied:
enum에 테스트를 추가하기가 애매하단 생각이 있었는데,
class와 동일하니 테스트를 추가하는게 좋은 것 같습니다.

chae-yh commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.Objects;
+
+public class Money {
+    private static final String DEFAULT_UNIT = "원";
```
이 부분도 view로 가는게 좋을 것 같습니다!

soob-forest replied:
view 계층으로 이동시켰습니다.

chae-yh commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,60 @@
+public class Money {
+    private static final String DEFAULT_UNIT = "원";
+    private final double value;
+    private final String unit;
```
현재 저희 기능 요건은 전부 동일 화폐를 기준으로 하고 있습니다. 예를 들어 달러와 원이 동시에 존재한다면 unit이 의미가 있겠지만, 현재 해당 필드의 필요성이 있을까요...?

soob-forest replied:
요구사항과 관련이 없는 것 같습니다.
제거했습니다.

chae-yh commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public Money add(Money target) {
+        return Money.from(this.value + target.value);
```
불변 객체 계산 좋습니다 : )

chae-yh commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    private void validate() {
+        if (this.value < 0) {
```
0은 매직넘버라 추출가능할 것 같습니다!

soob-forest replied:
추출했습니다.

chae-yh commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.List;
+import java.util.Set;
+
+public class WinningNumbers {
```
여기에서 lottoNumberSize가 필드로 필요할까요...?

soob-forest replied:
numberSize 제한이 없어도 될 것 같습니다.
제거했습니다.

chae-yh commented on src/test/java/lotto/domain/AutoLottoIssuerTest.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @Test
+    void 주문금액_예외() {
```
이러면 로또를 발행하지 못하는 것이 아니라, 총금액 / 로또 가격의 경우 가장 많이 구입할 수 있는 개수만큼 구입하면 될 것 같습니다!

soob-forest replied:
고민을 했던 부분인데, 해당 방향으로 수정했습니다.

chae-yh commented on src/test/java/lotto/domain/LottoStatisticTest.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+
+    private Lotto createLotto(int input0, int input1, int input2, int input3, int input4, int input5) {
```
6개의 파라미터가 아니라 리스트로 받으면 더 간편하지 않을까요...?

soob-forest replied:
코멘트 주신 방향으로 변경하겠습니다.

chae-yh commented on src/test/java/lotto/domain/LottosTest.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    @Test
+    void 당첨_로또_리스트() {
```
가능하면 문장 형태로 써주셔도 좋을 것 같아요. 정확히 어떤 의미인지 생각하게 되네요...

soob-forest replied:
네 알겠습니다.

chae-yh commented on src/test/java/lotto/domain/MoneyTest.java:
```diff
@@ -0,0 +1,37 @@
+    void setUp() {
+        money1 = Money.from(10000);
+        money2 = Money.from(1000);
+
```
소스를 보면 개행이 군데군데 개행이 많이 들어가 있는데 의미가 있는걸까요...?

soob-forest replied:
코드를 읽기 쉽게 하기위해, 코드 의미상 끝나는 부분에 개행을 삽입했는데, 필요한 부분에만 넣도록 해보겠습니다.

chae-yh commented on docs/lotto.md:
```diff
@@ -0,0 +1,39 @@
+
+### 로또 가격
+
+- [x] 1000원
```
요구사항을 조금 자세하게 적으면 어떨까요...?

soob-forest replied:
앗. 죄송합니다.. 
변경했습니다.

chae-yh commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.List;
+
+public class Lottos {
+    private final List<Lotto> lottoList;
```
변수명에 타입명은 가능한 쓰지 않는 편이 좋습니다! 예를 들어 이 자료구조를 Set으로 변경한다면 변수명을 전부 변경해주어야 합니다.

soob-forest replied:
lottos로 변경했습니다.

chae-yh commented on src/test/java/lotto/domain/LottoStatisticTest.java:
```diff
@@ -0,0 +1,68 @@
+
+        List<Lotto> lottoList = new ArrayList<>();
+        lottoList.add(createLotto(7, 8, 9, 19, 20, 21));
+        lottoList.add(createLotto(7, 8, 9, 19, 20, 21));
```
여기는 똑같은 값을 너무 많이 넣은것 같은데 사유가 있을까요...?

soob-forest replied:
로또 결과 통계 테스트를 위해, 당첨되지 않은 로또를 하드하게 반복적으로 넣었습니다...

최대한 정리 하는 방향으로 변경했습니다.

chae-yh reviewed (changes requested):
수배님 안녕하세요. 로또 미션 본격적 시작이시네요!
작성해주신 부분 중에 확인해보면 좋을 만한 부분에 대해서 코멘트 남겨드렸는데 확인 부탁드립니다 : )

파이팅 입니다!

soob-forest committed:
refactor(Lotto): 필요없는 멤버변수 제거

- 로또 넘버 사이즈를 static final 접근자로 변경
- 로또 가격을 static final 접근자로 변경

soob-forest committed:
refactor(Lotto): lottoNumbers의 자료구조를 더 효율적으로 변경함

- 중복이 불가능한 자료구조인 Set으로 변경

soob-forest committed:
refactor(Lotto): 멤버변수에 값을 할당하기 전에 validate하도록 변경

soob-forest committed:
refactor(MatchResult): matchCount를 이용한 매치결과 생성 함수 리팩토링

- if문을 이용해, 일일이 count를 비교하던 것에서, MatchResult.values()를 이용해 적절한 값을 찾도록 변경

- 예상치 못한 count값 일 경우 예외 핸들링

soob-forest committed:
refactor(AutoLottoIssuer): 주문 금액이 딱 맞아 떨어지지 않으면, 최대 금액만큼 주문하도록 변경

soob-forest committed:
refactor(Money): 안전한고 정확한 돈의 값 계산을 위해 BigDecimal으로 value 타입 변경

soob-forest committed:
refactor(LottoNumber): 값 할당 전에 validate하도록 변경

soob-forest committed:
refactor(LottoNumber): 로또 번호의 min, max를 상수로 변경

soob-forest committed:
refactor(LottoNumber): compareTo를 Integer.compare를 이용하도록 변경

soob-forest committed:
refactor(LottoRandomFactory): 불필요한 멤버변수 제거

soob-forest committed:
refactor(LottoStatistic): view 계층을 위한 코드 제거

soob-forest committed:
refactor(Lottos): matchedLottoList 반환 타입을 Lotto 일급 컬렉션으로 변경

soob-forest committed:
feat(Lottos): Lottos의 size 함수 추가

soob-forest committed:
refactor(LottoStatistic): 필요없는 멤버함수를 제거

- MatchResult를 key로 Lottos를 value로 가지는 멤버함수를 추가

soob-forest committed:
refactor(Money): 돈 최소값 상수로 선언

soob-forest committed:
refactor(Money): 요구사항에 존재하지 않는 필드 제거

soob-forest committed:
refactor(Money): 값을 할당하기 전에 validate를 하도록 변경

soob-forest committed:
refactor(WinningNumbers): 요구사항에 필요없는 멤버변수 제거

soob-forest committed:
refactor(WinningNumbers): 멤버변수를 적절한 자료구조로 변경

- 중복이 없는 데이터이기 때문에 Set으로 변경함

soob-forest committed:
refactor(WinningNumbers): validate 실행을 변수 대입전에 하도록 변경

soob-forest committed:
feat(MatchResult): 당첨인 MatchResult Enum 리스트를 반환 할 수 있도록 멤버함수 추가

soob-forest committed:
feat(LottoStatistic): 담첨 결과를 Set 형태로 반환하는 멤버함수 추가

soob-forest committed:
refactor(View): LottoStatistic 도메인에게 전달받은 데이터를 이용하도록 변경

soob-forest committed:
refactor(LottoController): 어플리케이션 클래스에 선언하고 사용했던 상수 제거

soob-forest committed:
refactor(LottoController): 각 함수가 하나의 기능을 하도록 리팩토링

soob-forest committed:
refactor(Lottos): 변수명에서 타입명 제거

soob-forest committed:
test: 테스트를 보기 좋게 변경

soob-forest committed:
docs: 요구사항 정리

chae-yh commented on src/main/java/lotto/domain/LottoRandomFactory.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    public Lotto create() {
+        LottoNumber[] lottoNumberList = new LottoNumber[LOTTO_NUMBER_SIZE];
```
여기를 Lotto.LOTTO_NUMBER_SIZE 로 쓸 수 있지 않을까요...? 그리고 만약 LOTTO_NUMBER_SIZE가 외부로 노출된다면 Lotto.SIZE 로만 써도 좋지 않을까요...?

chae-yh commented on src/main/java/lotto/domain/MatchResult.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public static MatchResult from(int matchCount) {
+        for (MatchResult matchResult : MatchResult.values()) {
```
저희가 indent depth를 2 이상 가져가지 않으려 하고 있습니다. 하지만 이런 경우는 indent를 1로 만들기 어렵더라구요.. 이런 경우에 한해서만 stream api를 사용해보는 것은 어떨까요...?

chae-yh commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public static Money from(double value) {
+        return new Money(new BigDecimal(Double.toString(value)));
```
BigDecimal.valueOf(value)로 해도 좋을 것 같습니다 : )

soob-forest commented:
테스트 코드 작성 시, 당첨 숫자 개수를 반환하는 함수 라던지, 통계를 내는 함수와 같은 것들의 테스트 작성시,
하드코딩이 요소가 많이 들어가게 되는 것 같고 코드가 지저분해 져서, 테스트 작성에 어려움이 있습니다. 
혹시 이를 잘 처리할 수 있는 방법이 있을지 궁금합니다.

soob-forest committed:
feat(Lotto): 정렬된 로또 번호 반환 함수 추가

soob-forest committed:
refactor(Lottos): Lotto class의 sortedLottoNumbers를 사용하도록 변경

soob-forest commented:
### Changed
- 요구사항 중, 정렬된 로또 번호를 반환하는 부분을 빠뜨렸었기 때문에, 해당 부분을 구현

리뷰를 하고 계실까봐, push하는것을 고민하다 commit 반영합니다.

chae-yh commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,59 @@
+    public List<LottoNumber> sortedLottoNumbers() {
+        List<LottoNumber> sortedNumbers = new ArrayList<>(this.lottoNumbers);
+        Collections.sort(sortedNumbers);
+        return sortedNumbers;
```
일급 컬렉션은 단순히 필드를 final로 선언하였다고 하여 불변객체가 되지 않습니다. lottoNumbers의 list를 재할당하는 것만을 막아줍니다. 해당 메소드에서 로또를 가져다가 사용자가 임의로 숫자를 넣어줄 수 있습니다. 이 부분을 막을 수 있도록 작성해보는 것도 좋을 것 같습니다. 다음과 같은 키워드로 공부해보시면 좋을 것 같아요!
Collections.unmodifiableList()

chae-yh commented on src/main/java/lotto/domain/LottoStatistic.java:
```diff
@@ -0,0 +1,72 @@
+        this(lottos, createWinningNumbers(winningNumbers));
+    }
+
+    private static WinningNumbers createWinningNumbers(String[] numbers) {
```
이 역할을 WinningNumbers에 보내면 어떨까요...?

chae-yh commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,50 @@
+    public Lottos matchedLottoList(WinningNumbers winningNumbers, MatchResult matchResult) {
+        List<Lotto> matchedLottos = new ArrayList<>();
+        for (Lotto lotto : lottos) {
+            if (winningNumbers.matchWinningLotto(lotto).equals(matchResult)) {
```
디미터 법칙(점을 여러개 찍지 마라!)을 어기는 것처럼 보이는데 이 부분도 역할을 적절히 분리해줄 수 있을 것 같습니다.
https://en.wikipedia.org/wiki/Law_of_Demeter

chae-yh commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,29 @@
+public class WinningNumbers {
+    private final Set<LottoNumber> winningNumbers;
+
+    public WinningNumbers(LottoNumber[] winningNumbers) {
```
이 부분은 단순 의견입니다만, 클래스명에 WinningLotto는 어떨까요...? Numbers보다는 직관적으로 이해하기 쉬울 것 같습니다!

chae-yh commented on src/test/java/lotto/domain/LottosTest.java:
```diff
@@ -0,0 +1,68 @@
+        lottoList.add(createLotto(new int[]{1, 2, 3, 4, 5, 6}));
+        lottoList.add(createLotto(new int[]{1, 2, 3, 4, 5, 7}));
+        lottoList.add(createLotto(new int[]{1, 2, 3, 4, 7, 8}));
+        lottoList.add(createLotto(new int[]{1, 2, 3, 7, 8, 9}));
```
질문 주신 내용 답변 여기서 드리겠습니다!
```
테스트 코드 작성 시, 당첨 숫자 개수를 반환하는 함수 라던지, 통계를 내는 함수와 같은 것들의 테스트 작성시,
하드코딩이 요소가 많이 들어가게 되는 것 같고 코드가 지저분해 져서, 테스트 작성에 어려움이 있습니다.
혹시 이를 잘 처리할 수 있는 방법이 있을지 궁금합니다.
```

이 부분을 제가 정확히 이해한 것인지는 모르겠습니다만, 테스트 코드 작성시 해당 부분과 같이 하드 코딩 요소가 너무 많다는 점 말씀해주신 것으로 이해했습니다!
이 경우 하드 코딩이 되는 객체, 예를 들어 LottoNumber, Lotto, Lottos 등의 객체를 미리 만들어두고 그 객체를 테스트 코드에서 사용하는 것은 어떨까요....?
이렇게 미리 만들어두는 객체를 Fixture라고 하는데, 해당 부분도 확인해보시면 좋을 것 같습니다!

혹시 제가 잘못 이해하고 있다거나 더 궁금한 부분이 있으시면 언제든 DM 등 질문 부탁드립니다.

chae-yh commented on src/main/java/lotto/domain/LottoStatistic.java:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    public BigDecimal lottoEarning() {
```
메소드 명이므로 동사구로 쓰면 좋을 것 같아요!

chae-yh reviewed (approved):
수배님 안녕하세요 피드백 잘 반영해주셨습니다 👍 이 정도면 이번 미션의 목표를 달성한 것으로 생각되어 merge 하겠습니다! 질문 주신 내용 및 몇가지 생각해보면 좋을 내용들은 코멘트로 남겨드렸는데 확인 부탁드립니다!

다음 미션 진행 부탁드립니다 감사합니다 : )
